### ISSUE TITLE
core dump when restarting redis while unable to resolve server name

### ISSUE THREAD

**Issue by brunoqc** [slact/nchan#188]
I don't have the dump yet since I'm testing it with docker and the alpine image and I'm not sure where the core dumps are generated (if they even are).

```
2016/05/25 18:08:21 [error] 7#7: REDISTORE: [Pubsub] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 18:08:21 [error] 7#7: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 18:08:21 [error] 7#7: REDISTORE: [Command] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 18:08:21 [error] 7#7: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 18:08:21 [error] 7#7: redis_nginx_adapter: could not create the redis context for some-redis:6379 - Non-recoverable error
2016/05/25 18:08:23 [alert] 1#1: worker process 7 exited on signal 11 (core dumped)
```

**Comment by brunoqc** [comment]
I figured out how to get core dumps but I wasn't able to reproduce the "Non-recoverable error" yet.

I was able to get a trace for a "Name does not resolve" error. I get this error if I shutdown the redis some-redis container.

```
2016/05/25 19:46:53 [error] 6#6: REDISTORE: [Pubsub] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 19:46:53 [error] 6#6: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 19:46:53 [error] 6#6: REDISTORE: [Command] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 19:46:53 [error] 6#6: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 19:46:54 [error] 6#6: redis_nginx_adapter: could not create the redis context for some-redis:6379 - Name does not resolve
2016/05/25 19:46:54 [alert] 1#1: worker process 6 exited on signal 11 (core dumped)
2016/05/25 19:46:54 [error] 18#18: redis_nginx_adapter: could not create the redis context for some-redis:6379 - Name does not resolve
2016/05/25 19:46:54 [alert] 1#1: worker process 18 exited on signal 11 (core dumped)
```

```
warning: Can't read pathname for load map: No error information.
Core was generated by `nginx: worker process'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  redisPublishCallback (c=0x55d7d5c9fa80, r=0x0, privdata=0x55d7d5c98640) at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:1621
1621      if(CHECK_REPLY_ARRAY_MIN_SIZE(reply, 2)) {
(gdb) bt
#0  redisPublishCallback (c=0x55d7d5c9fa80, r=0x0, privdata=0x55d7d5c98640) at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:1621
#1  0x000055d7d4a9f972 in __redisRunCallback (ac=ac@entry=0x55d7d5c9fa80, cb=cb@entry=0x7fff173682c0, reply=reply@entry=0x0) at /tmp/nchan-0.99.14/src/hiredis/async.c:269
#2  0x000055d7d4aa0284 in __redisAsyncFree (ac=ac@entry=0x55d7d5c9fa80) at /tmp/nchan-0.99.14/src/hiredis/async.c:283
#3  0x000055d7d4aa0469 in __redisAsyncDisconnect (ac=ac@entry=0x55d7d5c9fa80) at /tmp/nchan-0.99.14/src/hiredis/async.c:348
#4  0x000055d7d4aa04a8 in __redisAsyncHandleConnect (ac=ac@entry=0x55d7d5c9fa80) at /tmp/nchan-0.99.14/src/hiredis/async.c:503
#5  0x000055d7d4aa0bae in redisAsyncHandleWrite (ac=0x55d7d5c9fa80) at /tmp/nchan-0.99.14/src/hiredis/async.c:543
#6  0x000055d7d4aab082 in redis_nginx_write_event (ev=<optimized out>) at /tmp/nchan-0.99.14/src/store/redis/redis_nginx_adapter.c:152
#7  0x000055d7d4a0daeb in ngx_epoll_process_events (cycle=<optimized out>, timer=<optimized out>, flags=<optimized out>) at src/event/modules/ngx_epoll_module.c:929
#8  0x000055d7d4a0507f in ngx_process_events_and_timers (cycle=cycle@entry=0x55d7d5c8f9b0) at src/event/ngx_event.c:242
#9  0x000055d7d4a0bee7 in ngx_worker_process_cycle (cycle=cycle@entry=0x55d7d5c8f9b0, data=data@entry=0x0) at src/os/unix/ngx_process_cycle.c:753
#10 0x000055d7d4a0a53e in ngx_spawn_process (cycle=cycle@entry=0x55d7d5c8f9b0, proc=0x55d7d4a0be60 <ngx_worker_process_cycle>, data=0x0, name=0x55d7d4abbb65 "worker process", respawn=respawn@entry=0)
    at src/os/unix/ngx_process.c:198
#11 0x000055d7d4a0c915 in ngx_reap_children (cycle=0x55d7d5c8f9b0) at src/os/unix/ngx_process_cycle.c:621
#12 ngx_master_process_cycle (cycle=0x55d7d5c8f9b0) at src/os/unix/ngx_process_cycle.c:174
#13 0x000055d7d49e8fba in main (argc=<optimized out>, argv=<optimized out>) at src/core/nginx.c:367
```

**Comment by brunoqc** [comment]
When restarting the redis container. I got 2 dumps. One is the same as the last one.

```
2016/05/25 19:55:03 [error] 7#7: REDISTORE: gc_add chanhead /patate: already added
2016/05/25 19:55:03 [error] 8#8: REDISTORE: [Pubsub] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 19:55:03 [error] 8#8: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 19:55:03 [error] 8#8: REDISTORE: [Command] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 19:55:03 [error] 8#8: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 19:55:03 [error] 7#7: REDISTORE: [Pubsub] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 19:55:03 [error] 7#7: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 19:55:03 [error] 7#7: REDISTORE: [Command] connection to redis for some-redis:6379 closed: Server closed the connection
2016/05/25 19:55:03 [error] 7#7: redis_nginx_adapter: connection to redis failed - Server closed the connection
2016/05/25 19:55:03 [error] 7#7: redis_nginx_adapter: could not select redis database
2016/05/25 19:55:03 [error] 7#7: REDISTORE: [Pubsub] connection to redis for some-redis:6379 closed: Connection refused
2016/05/25 19:55:03 [error] 7#7: redis_nginx_adapter: connection to redis failed - Connection refused
2016/05/25 19:55:03 [error] 7#7: redis_nginx_adapter: could not select redis database
2016/05/25 19:55:04 [alert] 1#1: worker process 7 exited on signal 11 (core dumped)
2016/05/25 19:55:04 [error] 31#31: redis_nginx_adapter: could not create the redis context for some-redis:6379 - Name does not resolve
2016/05/25 19:55:04 [alert] 1#1: worker process 31 exited on signal 11 (core dumped)
2016/05/25 19:55:04 [error] 32#32: redis_nginx_adapter: could not create the redis sync context for some-redis:6379 - Connection refused
2016/05/25 19:55:05 [error] 32#32: IPC:got IPC alert delayed by 2 sec
2016/05/25 19:55:05 [error] 32#32: redis_nginx_adapter: could not select redis database
2016/05/25 19:55:05 [error] 32#32: REDISTORE: [Command] connection to redis for some-redis:6379 closed: Connection refused
2016/05/25 19:55:05 [error] 32#32: redis_nginx_adapter: connection to redis failed - Connection refused
2016/05/25 19:55:05 [error] 32#32: redis_nginx_adapter: could not select redis database
2016/05/25 19:55:05 [error] 32#32: REDISTORE: [Pubsub] connection to redis for some-redis:6379 closed: Connection refused
2016/05/25 19:55:05 [error] 32#32: redis_nginx_adapter: connection to redis failed - Connection refused
```

```
warning: Can't read pathname for load map: No error information.
Core was generated by `nginx: master process'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __redisAsyncCommand (ac=ac@entry=0x0, fn=fn@entry=0x559edef57d90 <redis_load_script_callback>, privdata=privdata@entry=0x559edf1cad00 <redis_lua_scripts>,
    cmd=0x559ee0e71e80 "*3\r\n$6\r\nSCRIPT\r\n$4\r\nLOAD\r\n$585\r\n--input:  keys: [], values: [channel_id, number]\n--output: current_fake_subscribers\n  \nredis.call('echo', ' ####### FAKESUBS ####### ')\nlocal id=ARGV[1]\nlocal num=tonum"..., len=619) at /tmp/nchan-0.99.14/src/hiredis/async.c:594
594         if (c->flags & (REDIS_DISCONNECTING | REDIS_FREEING)) return REDIS_ERR;
(gdb) bt
#0  __redisAsyncCommand (ac=ac@entry=0x0, fn=fn@entry=0x559edef57d90 <redis_load_script_callback>, privdata=privdata@entry=0x559edf1cad00 <redis_lua_scripts>,
    cmd=0x559ee0e71e80 "*3\r\n$6\r\nSCRIPT\r\n$4\r\nLOAD\r\n$585\r\n--input:  keys: [], values: [channel_id, number]\n--output: current_fake_subscribers\n  \nredis.call('echo', ' ####### FAKESUBS ####### ')\nlocal id=ARGV[1]\nlocal num=tonum"..., len=619) at /tmp/nchan-0.99.14/src/hiredis/async.c:594
#1  0x0000559edef51c82 in redisvAsyncCommand (ac=0x0, fn=0x559edef57d90 <redis_load_script_callback>, privdata=0x559edf1cad00 <redis_lua_scripts>, format=<optimized out>, ap=ap@entry=0x7ffd55330800)
    at /tmp/nchan-0.99.14/src/hiredis/async.c:660
#2  0x0000559edef51d3e in redisAsyncCommand (ac=ac@entry=0x0, fn=fn@entry=0x559edef57d90 <redis_load_script_callback>, privdata=privdata@entry=0x559edf1cad00 <redis_lua_scripts>,
    format=format@entry=0x559edef85ec6 "SCRIPT LOAD %s") at /tmp/nchan-0.99.14/src/hiredis/async.c:669
#3  0x0000559edef57fff in redisInitScripts (c=0x0) at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:376
#4  rds_ctx () at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:385
#5  0x0000559edef5aeb1 in nchan_store_init_worker (cycle=<optimized out>) at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:276
#6  0x0000559edef662d4 in nchan_init_worker (cycle=0x559ee0e729b0) at /tmp/nchan-0.99.14/src/nchan_setup.c:37
#7  0x0000559edeebc936 in ngx_worker_process_init (cycle=cycle@entry=0x559ee0e729b0, worker=worker@entry=0) at src/os/unix/ngx_process_cycle.c:900
#8  0x0000559edeebce80 in ngx_worker_process_cycle (cycle=cycle@entry=0x559ee0e729b0, data=data@entry=0x0) at src/os/unix/ngx_process_cycle.c:734
#9  0x0000559edeebb53e in ngx_spawn_process (cycle=cycle@entry=0x559ee0e729b0, proc=0x559edeebce60 <ngx_worker_process_cycle>, data=0x0, name=0x559edef6cb65 "worker process", respawn=respawn@entry=0)
    at src/os/unix/ngx_process.c:198
#10 0x0000559edeebd915 in ngx_reap_children (cycle=0x559ee0e729b0) at src/os/unix/ngx_process_cycle.c:621
#11 ngx_master_process_cycle (cycle=0x559ee0e729b0) at src/os/unix/ngx_process_cycle.c:174
#12 0x0000559edee99fba in main (argc=<optimized out>, argv=<optimized out>) at src/core/nginx.c:367
```

```
Core was generated by `nginx: worker process'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  redisPublishCallback (c=0x559ee0e7bd00, r=0x0, privdata=0x559ee0e8bbe0) at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:1621
1621      if(CHECK_REPLY_ARRAY_MIN_SIZE(reply, 2)) {
(gdb) bt
#0  redisPublishCallback (c=0x559ee0e7bd00, r=0x0, privdata=0x559ee0e8bbe0) at /tmp/nchan-0.99.14/src/store/redis/rdsstore.c:1621
#1  0x0000559edef50972 in __redisRunCallback (ac=ac@entry=0x559ee0e7bd00, cb=cb@entry=0x7ffd553308b0, reply=reply@entry=0x0) at /tmp/nchan-0.99.14/src/hiredis/async.c:269
#2  0x0000559edef51284 in __redisAsyncFree (ac=ac@entry=0x559ee0e7bd00) at /tmp/nchan-0.99.14/src/hiredis/async.c:283
#3  0x0000559edef51469 in __redisAsyncDisconnect (ac=ac@entry=0x559ee0e7bd00) at /tmp/nchan-0.99.14/src/hiredis/async.c:348
#4  0x0000559edef514a8 in __redisAsyncHandleConnect (ac=ac@entry=0x559ee0e7bd00) at /tmp/nchan-0.99.14/src/hiredis/async.c:503
#5  0x0000559edef51bae in redisAsyncHandleWrite (ac=0x559ee0e7bd00) at /tmp/nchan-0.99.14/src/hiredis/async.c:543
#6  0x0000559edef5c082 in redis_nginx_write_event (ev=<optimized out>) at /tmp/nchan-0.99.14/src/store/redis/redis_nginx_adapter.c:152
#7  0x0000559edeebeaeb in ngx_epoll_process_events (cycle=<optimized out>, timer=<optimized out>, flags=<optimized out>) at src/event/modules/ngx_epoll_module.c:929
#8  0x0000559edeeb607f in ngx_process_events_and_timers (cycle=cycle@entry=0x559ee0e729b0) at src/event/ngx_event.c:242
#9  0x0000559edeebcee7 in ngx_worker_process_cycle (cycle=cycle@entry=0x559ee0e729b0, data=data@entry=0x0) at src/os/unix/ngx_process_cycle.c:753
#10 0x0000559edeebb53e in ngx_spawn_process (cycle=cycle@entry=0x559ee0e729b0, proc=proc@entry=0x559edeebce60 <ngx_worker_process_cycle>, data=data@entry=0x0,
    name=name@entry=0x559edef6cb65 "worker process", respawn=respawn@entry=-3) at src/os/unix/ngx_process.c:198
#11 0x0000559edeebc15c in ngx_start_worker_processes (cycle=cycle@entry=0x559ee0e729b0, n=2, type=type@entry=-3) at src/os/unix/ngx_process_cycle.c:358
#12 0x0000559edeebd652 in ngx_master_process_cycle (cycle=0x559ee0e729b0) at src/os/unix/ngx_process_cycle.c:130
#13 0x0000559edee99fba in main (argc=<optimized out>, argv=<optimized out>) at src/core/nginx.c:367
```

**Comment by slact** [comment]
Ok, I see what's happening here. I'll have it fixed shortly.

**Comment by slact** [comment]
fixed in 352b57d , I'll test it a little more before merging into master.

**Comment by slact** [comment]
merged into master and released in v0.99.15

**Comment by brunoqc** [comment]
Looks good! :+1:
